### DISCUSSION TITLE
macOS catalina

### DISCUSSION THREAD

**Discussion by aldamepi** [Homebrew/discussions#3851] [Everyday usage]
brew is no longer working on macOS 10.15 Catalina, right?

**Comment by Bo98** [comment]
It "works" but we will soon stop updating pre-built bottles, so when you install formulae it may build them from source instead which can be very slow and prone to breaking.

**Reply by aaronliu0130** [reply]
Hi, is there a discussion where the decision to stop updating catlalina bottles were made?

**Reply by SMillerDev** [reply]
We only support the versions that Apple supports, which is at most 2 versions back. There's no specific discussion for Catalina.

**Reply by jamesquilty** [reply]
It would be helpful for those of us who need stability to be able to set `HOMEBREW_NO_AUTO_UPDATE=1`  and `git reset --hard <commit-sha>` on the local clones to prevent Homebrew from breaking our working systems and costing us several hours of time fixing them (often for the very marginal benefit of minor version updates). I searched but couldn't find anything immediately: is there a specific Issue or git tag which marks the transition from "supported" to "not supported"?

**Reply by aaronliu0130** [reply]
Both of these are supported? for the first one did you forget to export it? for the second one just navigate to the $(brew --repository) and find the tap in the library taps homebrew homebrew-core directory. Plus homebrew's auto update won't auto upgrade your packages

**Reply by p-linnane** [reply]
The transition is marked by Apple releasing a new OS. When Ventura was released, Catalina became unsupported. It took us a bit to get Ventura CI all set up, so Catalina bottles were still being built for about a month after it was released.

**Reply by aaronliu0130** [reply]
btw I remember somewhere in the docs discussing that but i can't find it anymore, may somebody kindly link me to it?

**Reply by p-linnane** [reply]
https://docs.brew.sh/Installation#2

**Reply by aaronliu0130** [reply]
thanks but i mean the "last two major releases" part specifically

**Reply by p-linnane** [reply]
I don't know if there is something that specifically calls it out in those terms, but we update the requirements upon new macOS releases. There is also [this](https://github.com/Homebrew/homebrew-core/pull/118642) which indicates formulae must build on the 3 latest macOS releases & Linux to be included.

**Reply by jamesquilty** [reply]
@p-linnane It seems a bit... _odd_ that something external to Homebrew implicitly marks an important internal change. Perhaps this is something that could be addressed by making the change explicitly within Homebrew?

@aaronliu0130 You write "find the tap" but the question is _actually_ "which commit?". Users will upgrade periodically &mdash; it's why we want a package manager &mdash; the problem is that `brew outdated` mentions nothing about OS support _but_ `brew upgrade` warns about out-of-support _after_ the upgrade process is underway. I'd prefer to know beforehand that Homebrew is going to break my system, if that can't happen then I'd like to be able to back-out to the previous known working state &mdash; a basic function of most package managers &mdash; which in Homebrew's case means knowing commit SHAs.

Just to be crystal clear: there was no indication that on this particular day `brew upgrade` was going to break my system. Seems like a problem that should be fixed, at least to me.

**Reply by aaronliu0130** [reply]
1. again i think there was a place on the website that said it. 2. how does a brew upgrade, where most of the packages that don't have updated bottles can't be installed, break your system? i don't remember any pack mans i've used to return to working state

**Reply by p-linnane** [reply]
I'm not sure that it's so 'external' to Homebrew. We only support the OS's Apple supports. Here is the commit marking Ventura as supported and Catalina as unsupported: https://github.com/Homebrew/brew/commit/e16afb305e8ffd493554c4b6d9ba48764f4b166b

If you are using macOS Catalina (released over 3 years ago now) here at the end of 2022, I'm not sure what to tell you. You've probably been getting OS update prompts for 2 years. It's the same as people disabling Gatekeeper. If you choose not to stay up to date, or willingly disable security features, we can only warn you that it's not a good idea, and unsupported.

We don't actively try to break things on older systems ([it's quite the opposite actually](https://github.com/Homebrew/homebrew-core/pull/85775#issuecomment-995361515)), but we can't spend our limited time as maintainers on these older sytems.

**Reply by jamesquilty** [reply]
@p-linnane Thank you, the commit SHA is helpful; you don't have to say any more than that. It would be appropriate, I think, for there to be prior warning of expected breakage so that I (and others) don't have to spend my limited time fixing packages that break. Specific suggestions:

- `brew outdated` displays the `Warning: You are using macOS X.Y` message and/or
- `brew upgrade` displays an interactive message on the console with the "expected breakage" warning and asks if the user wishes to proceed [current behaviour is to warn that breakage is expected and then to proceed anyway!].

That seems fair to me, I hope it seems the same way to you.

**Reply by SMillerDev** [reply]
> brew outdated displays the Warning: You are using macOS X.Y message

Apple already does that in system settings. No need to duplicate it across all brew commands.

> brew upgrade displays an interactive message

Brew never displays interactive messages.

---

If you don't want to keep your system up to date you're in trouble for many reasons and many applications can suddenly break. Homebrew isn't special in this so I don't see a need to cater heavily for that use case.

**Reply by jamesquilty** [reply]
> Apple already does that in system settings. No need to duplicate it across all brew commands.

 Excuse me, I thought it was clear in context that I was referring to the _entire_ message that `brew update` displays. If the message (appended) is important enough to display by one command then that suggests the message should also be displayed by other brew commands.

> Brew never displays interactive messages.

 All the more reason to display the warning message in commands such as `brew outdated`.

> Homebrew isn't special in this so I don't see a need to cater heavily for that use case.

 I'm asking for a small consideration: a change to the way Homebrew displays messages warning of expected breakage. I consider my time to be at least as valuable as yours and having prior warning would have saved me a few hours. I wouldn't consider adding the warning message to other commands to be a "heavy catering" when a warning is already given at the time changes are made (too late!).

 This would improve the user-friendliness of Homebrew, isn't that something you want?

```text
Warning: You are using macOS 10.15.
We (and Apple) do not provide support for this old version.
It is expected behaviour that some formulae will fail to build in this old version.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from MacHomebrew on Twitter.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

**Reply by SMillerDev** [reply]
> I consider my time to be at least as valuable as yours

Within the context of Homebrew, there would be no Homebrew without maintainers. There would absolutely still be a Homebrew without users that don't maintain.

Let's put it like this, if you make a pull request I'm sure people will happily review it. Any other effort from Homebrew maintainers to convince people, who ignore their systems update notifications for 2 years, to update is wasted in my book.

**Reply by jamesquilty** [reply]
Convincing people to update is not the matter under discussion.

**Reply by SMillerDev** [reply]
Then I'm even less sure what you're hoping to accomplish with more messages.

**Reply by jamesquilty** [reply]
@SMillerDev: I was hoping you'd address the matter of displaying warnings, which I've been very consistent in raising. Could you do that in a respectful way please?

**Reply by SMillerDev** [reply]
I see you are consistently raising this issue, and I am consistently telling you that is is time that would have to be invested by you because it is requested by you. Maintainers, although free to do whatever they want, should in my opinion not be wasting time on improving unsupported use cases when they could be improving supported ones.

**Reply by aaronliu0130** [reply]
He means that if you fork and make a pull request that adds that warning the maintainers would be happy to review it but they won't open a pull request for you.

**Reply by euguk007** [reply]
Homebrew is dead for anything below Big Sur. This means it is dead for most developers using macOS and below. Basically, Homebrew is a dieing product.

**Reply by aaronliu0130** [reply]
You rely on the assumption that the majority of developers are using catalina. That is not true.

**Reply by euguk007** [reply]
Ha ha. Did you know that some use Mavericks?
The point is that Homebrew could be more considerate towards those who have older hardware.

**Reply by p-linnane** [reply]
You can use whatever you want. We will not support it, and will continue to only support the operating systems Apple supports.

This fall when macOS 14 comes out, Big Sur will become unsupported. And that cadence will continue every year.

**Reply by euguk007** [reply]
@p-linnane You just proved my point. Homebrew is dead for me and many others.

**Reply by aaronliu0130** [reply]
> Did you know that some use Mavericks?

Did you know that some use Windows ME?
Point is, not supporting hardware that has been out for over 5 years does not kill a product.

**Reply by aaronliu0130** [reply]
"Many others" needs a source. There are no current accurate surveys for userbase division among the last 4 releases because the source cannot differentiate between catalina and higher versions.

**Reply by euguk007** [reply]
@aaronliu0130 - The future is with the flexible kind. Not with the stiff ones ;)

**Reply by aaronliu0130** [reply]
Flexible ones are able to get new computers after more than 5 years. For example my machine that was "decommissioned" by this change is an 2014 macbook air and several others that only have 128 gb storage.

**Reply by euguk007** [reply]
@aaronliu0130 - mate. i think you are using this platform as FB. making a pointless conversation.
I made my point and I am out.

**Comment by jamesquilty** [comment]


**Reply by aaronliu0130** [reply]
Since this isn't a new answer it should just be a reply.
